### PR TITLE
feat(logs): add --min-level severity threshold filter

### LINKED ISSUE #523: Add `azd app logs --min-level` to filter by severity threshold
## Problem

`azd app logs --level` matches one exact severity, so there's no way to say "show me warnings and anything worse" in a single command. To watch warnings plus errors today you run the command twice and merge the output by eye.

## Solution

Add `--min-level <level>` that shows entries at the given severity and above, using the order debug < info < warn < error. `--min-level warn` prints warn and error entries; `--min-level info` drops debug. Keep `--level` as the exact-match filter. The two flags are mutually exclusive, and `--min-level` can't be combined with `--context` (which is tied to an exact level).

## Acceptance criteria

- [ ] `--min-level` accepts debug, info, warn, error (not `all`) and errors on anything else.
- [ ] Entries are kept when their severity is at or above the threshold, using debug < info < warn < error.
- [ ] Passing both `--level` (non-`all`) and `--min-level`, or `--min-level` with `--context`, returns a clear validation error.
- [ ] Works in both the default and `--follow` paths.
- [ ] Table-driven unit tests cover the ordering, the boundary cases, and the flag conflicts.
- [ ] `--help`, `cli/docs/cli-reference.md`, and `cli/docs/commands/logs.md` document the flag.

### PR DESCRIPTION
**Pull request by jongio**
## What

Adds a `--min-level` flag to `azd app logs` that shows entries at a given severity or higher.

Severity order is `debug < info < warn < error`. For example, `--min-level warn` shows warn and error entries and hides debug and info.

## Why

`--level` matches a single severity exactly. When you are chasing a problem you usually want "warnings and worse" without listing each level. `--min-level` gives you that threshold in one flag.

## Details

- New `--min-level` flag on `azd app logs`. Accepts `debug`, `info`, `warn` (or `warning`), `error`.
- Works in both batch mode and follow mode (`--follow`).
- Cannot be combined with `--level` or `--context`; the command returns a clear error if you try.
- Severity is mapped explicitly so the raw `LogLevel` iota order does not affect ordering.

## Tests

- `TestParseMinLevel`, `TestLogLevelSeverity`, `TestMeetsMinLevel`, `TestFilterLogsByMinLevel`, and `TestValidateLogsOptionsMinLevel` cover parsing, ordering, filtering, and validation including the mutual-exclusion errors.
- `go build ./...`, the commands package tests, and `golangci-lint` all pass.

Closes #523

### CODE REVIEW

**Review by wbreza — COMMENTED** [review]
Solid, well-tested severity-threshold filter; two items to address — the help/docs overstate the --level exclusion, and the runtime wiring lacks an execution-path test.

### What stood out
- **Correct severity model** — Explicit debug < info < warn < error mapping decouples ordering from the LogLevel iota (Info=0, Warn=1, Error=2, Debug=3), avoiding a real mis-ranking bug.
- **Consistent two-path wiring** — --min-level is applied once per entry in both batch collect() and follow shouldDisplayEntry, and validation prevents overlap with --level/--context.

### Top issues
- [F-001] Help/docs claim "cannot be combined with --level" but `--level all` is accepted — `cli/src/cmd/app/commands/logs.go:301`
- [F-002] --min-level runtime wiring has no execution-path test — `cli/src/cmd/app/commands/logs.go:582`

### Improvement opportunities
- [O-001] Parse --min-level once per follow session, not per streamed entry _(Recommended)_ — `cli/src/cmd/app/commands/logs_filtering.go:177`
- [O-002] Match filterLogsByMinLevel capacity heuristic to filterLogsByLevel _(Optional)_ — `cli/src/cmd/app/commands/logs_filtering.go:273`

File: cli/src/cmd/app/commands/logs.go [hunk 1]
```diff
@@ -294,6 +298,7 @@ Examples:
 	cmd.Flags().BoolVar(&opts.timestamps, "timestamps", true, "Show timestamps with each log entry")
 	cmd.Flags().BoolVar(&opts.noColor, "no-color", false, "Disable colored output")
 	cmd.Flags().StringVar(&opts.level, "level", "all", "Filter by log level (info, warn, error, debug, all)")
+	cmd.Flags().StringVar(&opts.minLevel, "min-level", "", "Show entries at this severity or higher (debug < info < warn < error); cannot be combined with --level or --context")
```
Comment (wbreza) [inline cli/src/cmd/app/commands/logs.go]: [F-001] The flag help and docs state --min-level "cannot be combined with --level," but validateLogsOptions only rejects the combination when --level is set to a non-default value (`strings.ToLower(opts.level) != "all"` at logs_filtering.go:356). Because --level defaults to `all`, `azd app logs --level all --min-level warn` is accepted and silently applies the threshold, contradicting the stated constraint. Permitting the default `all` is required for --min-level to work at all, so the correct fix is to align the wording here and in the docs tables (cli/docs/commands/logs.md, cli/docs/cli-reference.md) to say --min-level cannot be combined with an explicit --level (info/warn/error/debug) or --context.

```suggestion
	cmd.Flags().StringVar(&opts.minLevel, "min-level", "", "Show entries at this severity or higher (debug < info < warn < error); cannot be combined with an explicit --level (info/warn/error/debug) or --context")
```

File: cli/src/cmd/app/commands/logs.go [hunk 2]
```diff
@@ -572,6 +577,12 @@ func (e *logsExecutor) collect(ctx context.Context, args []string) (*CollectedLo
 		// Regular mode: filter by level
 		logs = filterLogsByLevel(logs, levelFilter)
 
+		// Filter by minimum severity threshold when --min-level is set.
+		// Mutually exclusive with --level, so only one of these actually filters.
+		if minLevel, ok := parseMinLevel(e.opts.minLevel); ok {
```
Comment (wbreza) [inline cli/src/cmd/app/commands/logs.go:582]: [F-002] The new --min-level behavior is wired into two runtime paths — batch mode here in collect() and follow mode in shouldDisplayEntry (logs_filtering.go:177) — but no test drives either path with opts.minLevel set. The five new tests cover only the pure helpers (parseMinLevel, logLevelSeverity, meetsMinLevel, filterLogsByMinLevel) and validateLogsOptions; deleting both call sites (logs.go:582-584 and logs_filtering.go:177-181) would leave every test green, so the user-visible filtering is unverified. Add an execution-path test: drive shouldDisplayEntry with e.opts.minLevel="warn" (levelFilter=LogLevelAll) asserting debug/info are dropped and warn/error kept, plus a collect()-level test over a synthetic slice asserting the returned Entries are threshold-filtered.

File: cli/src/cmd/app/commands/logs_filtering.go [hunk 3]
```diff
@@ -173,6 +173,13 @@ func (e *logsExecutor) shouldDisplayEntry(entry service.LogEntry, levelFilter se
 		return false
 	}
 
+	// Filter by minimum severity threshold when --min-level is set
+	if minLevel, ok := parseMinLevel(e.opts.minLevel); ok {
```
Comment (wbreza) [inline cli/src/cmd/app/commands/logs_filtering.go]: 💡 Improvement Opportunity (optional — not required to merge)
**Benefit:** Removes a per-entry strings.ToLower + switch from the follow-mode hot loop; shouldDisplayEntry runs parseMinLevel(e.opts.minLevel) for every streamed entry.
**Effort:** S
**Optionality:** Recommended
**Sketch:** Parse the threshold once at the start of the follow session (mirroring how levelFilter is already resolved) and pass the resolved service.LogLevel plus a "set" bool into shouldDisplayEntry instead of re-parsing the raw string on every call.

File: cli/src/cmd/app/commands/logs_filtering.go [hunk 4]
```diff
@@ -220,6 +227,58 @@ func filterLogsByLevel(logs []service.LogEntry, level service.LogLevel) []servic
...
+}
+
+// meetsMinLevel reports whether an entry's level is at or above the given threshold.
+func meetsMinLevel(entryLevel, minLevel service.LogLevel) bool {
+	return logLevelSeverity(entryLevel) >= logLevelSeverity(minLevel)
+}
+
+// filterLogsByMinLevel keeps only entries at or above the given severity threshold.
+func filterLogsByMinLevel(logs []service.LogEntry, minLevel service.LogLevel) []service.LogEntry {
+	filtered := make([]service.LogEntry, 0, len(logs))
```
Comment (wbreza) [inline cli/src/cmd/app/commands/logs_filtering.go]: 💡 Improvement Opportunity (optional — not required to merge)
**Benefit:** Keeps the two level filters symmetric and avoids over-allocating when a high threshold discards most entries.
**Effort:** S
**Optionality:** Optional
**Sketch:** filterLogsByMinLevel pre-allocates len(logs), while filterLogsByLevel uses len(logs)/filterCapacityEstimate. Applying the same heuristic here prevents a large batch filtered by --min-level error from over-allocating.

**Review by wbreza — APPROVED** [review]
Re-review of the update (through 422243a). All four items from the prior round are resolved, and the accompanying refactor is regression-clean. Approving.

### What stood out
- **All prior feedback addressed** — The Medium wording mismatch and the Low test-coverage gap are both fixed, and both optional opportunities were taken.
- **Disciplined signature refactor** — Threading pre-parsed `minLevel, minLevelSet` through the entire follow chain was done with correct positional order at every call site; build and the full command-package test suite pass.

### Resolved since last round
- [F-001] `--min-level` help, docs (logs.md, cli-reference.md), the validation error, and its test now consistently state the constraint is against an explicit `--level (info/warn/error/debug)`, matching that `--level all` is permitted — `cli/src/cmd/app/commands/logs.go:301`, `cli/src/cmd/app/commands/logs_filtering.go:359`.
- [F-002] Execution-path tests added: `TestLogsExecutor_CollectFiltersByMinLevel` (batch) and `TestShouldDisplayEntryMinLevel` (follow) drive the real wiring and assert filtered output.
- [O-001] `parseMinLevel` is now called once per path (`logs_streaming.go:16`, `logs.go:582`); the per-entry parse in `shouldDisplayEntry` is gone.
- [O-002] `filterLogsByMinLevel` now uses the same `len(logs)/filterCapacityEstimate` capacity heuristic (floor 10) as `filterLogsByLevel`.

### Verified this round
- Positional argument order is correct across all follow functions and every `shouldDisplayEntry` call site (no `levelFilter`/`minLevel` swap despite both being `service.LogLevel`).
- No path double-applies the threshold; follow relies only on `shouldDisplayEntry`, batch only on `filterLogsByMinLevel`.
- `go build ./...` clean and the full `cmd/app/commands` package test suite passes.

File: cli/src/cmd/app/commands/logs.go (hunk 1 above)
Comment (jongio) [reply cli/src/cmd/app/commands/logs.go]: Fixed in [422243a3](https://github.com/jongio/azd-app/commit/422243a3). You are right that permitting the default `all` is exactly what makes `--min-level` usable, so I aligned the wording to the behavior rather than tightening the validation. The flag help and both docs tables now say `--min-level` cannot be combined with an explicit `--level` or with `--context`.

File: cli/src/cmd/app/commands/logs.go (hunk 2 above)
Comment (jongio) [reply cli/src/cmd/app/commands/logs.go:582]: Fixed in [422243a3](https://github.com/jongio/azd-app/commit/422243a3). Added an execution-path test driving `shouldDisplayEntry` in follow mode, plus a `collect()` test over a synthetic slice. I verified both are real by deleting each call site in turn and confirming the matching test fails, rather than assuming coverage from the line numbers.

File: cli/src/cmd/app/commands/logs_filtering.go (hunk 3 above)
Comment (jongio) [reply cli/src/cmd/app/commands/logs_filtering.go]: Fixed in [422243a3](https://github.com/jongio/azd-app/commit/422243a3). Follow mode now parses the threshold once in `followLogs` and passes the resolved level plus a set bool through the streaming paths, so the hot loop no longer re-parses per entry.

File: cli/src/cmd/app/commands/logs_filtering.go (hunk 4 above)
Comment (jongio) [reply cli/src/cmd/app/commands/logs_filtering.go]: Fixed in [422243a3](https://github.com/jongio/azd-app/commit/422243a3). `filterLogsByMinLevel` now uses the same `len(logs) / filterCapacityEstimate` heuristic as `filterLogsByLevel`, with a small floor so tiny inputs do not allocate a zero-capacity slice.
